module1/homework/make_unique/make_unique-ut.cpp

```diff
@@ -0,0 +1,67 @@
+#include <gtest/gtest.h>
+#include <memory>
+#include "make_unique.hpp"
+
+constexpr int testIntValue = 10;
+constexpr double testDoubleValue = 3.14;
+constexpr bool testBoolValue = true;
+constexpr size_t arraySize = 5;
+
+template<class Type> struct default_delete;
+template<class Type> struct default_delete<Type[]>;
+
+struct make_uniqueTest {
+public:
+    make_uniqueTest();
+    make_uniqueTest(int intValue, double doubleValue, bool boolValue)
+    : intValue_(intValue), doubleValue_(doubleValue), boolValue_(boolValue)
+    {}
+
+    bool operator==(const make_uniqueTest& other) const {
+        return intValue_ == other.intValue_
+            && doubleValue_ == other.doubleValue_
+            && boolValue_ == other.boolValue_;
+    }
+
+private:
+    int intValue_;
+    double doubleValue_;
+    bool boolValue_;
+};
+
+TEST(make_uniqueIntTest, testMakeUniqueForIntValue) {
+    auto uPtr = cs::make_unique<int>(testIntValue);
+    ASSERT_EQ(*uPtr, testIntValue);
+}
+
+TEST(make_uniqueArrayTest, testMakeUniqueForArrayType) {
+    auto uArray = cs::make_unique<int[]>(arraySize);
+    for (size_t i =0; i < arraySize; i++) {
+        uArray[i] = i;
+    }
+    for (size_t i =0; i < arraySize; i++) {
+        ASSERT_EQ(uArray[i], i);
+    }
+}
+
+TEST(make_uniqueEmptyTest, testIntEmptyConstruct) {
+    auto uniqZero = cs::make_unique<int>();
+    ASSERT_EQ(*uniqZero, 0);
+}
+
+TEST(make_uniqueLvalueTest, testLvalueConstruct) {
+    int lvalue{2020};
+    auto uniqTwenty = cs::make_unique<int>(lvalue);
+    ASSERT_EQ(*uniqTwenty, lvalue);
+}
+
+TEST(make_uniqueCustomObjectTest, testLvalueDoubleBoolConstruct) {
+    make_uniqueTest muTest{testIntValue, testDoubleValue, testBoolValue};
+    auto uniqeObject = cs::make_unique<make_uniqueTest>(testIntValue, testDoubleValue, testBoolValue);
+    ASSERT_EQ(*uniqeObject, muTest);
+
+    int lvalue{2020};
+    make_uniqueTest muTest2{lvalue, testDoubleValue, testBoolValue};
+    auto uniqeObject2 = cs::make_unique<make_uniqueTest>(lvalue, testDoubleValue, testBoolValue);
+    ASSERT_EQ(*uniqeObject2, muTest2);
+}
```

Comment:
You have test for various number of arguments and for arrays. But r-values are not tested. lvalue, testDoubleValue and testBoolValue are all l-values.

To test it you need your own structure with copy and move semantics. Then you should check if there was a move constructor / assignment operator called when passing rvalue and copy constructor / assignment operator when passing lvalue.

Reply:
done with move and assignment constructors

Reply:
OK, but the question is - do you know which constructor was called in each example? How can you be sure that it was the proper one? How can you check which one was called? :)
Your test cannot tell it.

Reply:
so i can just simply force the exact constructor with std::move

Reply:
no, std::move doesn't guarantee that move construcor will be called. If it does not exist copy constructor is called.

You can change constructors behavior to set some predefined values in each case or add a new field inside your test object that will be set differently in each constructor.

Reply:
right std::move doesn't move :) in my case constructors were used correctly, added fourth attribute to second constructor.
Now I'm sure which is used in test case